### ISSUE TITLE
Memory Limits

### ISSUE THREAD

**Issue by bdrodes** [celery/celery#5645]
Is it possible to add a capability to celery to specify a max amount of memory for an entire blade/machine? 

The current flag (--max-memory-per-child) appears to be a memory limit per child task, but ideally I would like to specify a total memory consumption for all tasks on the same machine. That is, I want a limit on the total amount of memory all tasks currently executing on a single machine may collectively consume. If that limit is reached, I don't want any further tasks to be dispatched on that machine regardless of concurrency settings. Is this a capability of --max-memory-per-child and the documentation is just limited, or is this a new feature request?

**Comment by clokep** [comment]
You should roughly be able to take (number of worker threads * max memory per child) and that will give you the expected maximum memory in use, roughly. (This leaves out the memory in use by the parent process.)

I'm a bit confused at what you're trying to accomplish though.

**Comment by bdrodes** [comment]
I have two related issues with memory limits I'm trying to solve:

1) Individual task memory limits. Once a limit is reached, I want celery to send a termination signal to the task. Right now, there appears to be no support for that (which I'll open a new issue for that unless you can provide some pointers here). To achieve this I have to do a resource.setrlimit in my tasks.

2) collective task memory limits. An individual task might not consume a lot of memory, but if I have too many running in parallel, the total memory consumption my Out Of Memory my machine. I'd rather not limit the number of cpus that can be used, but instead have celery observe it is over scheduling by monitoring memory. I want to be able to specify to celery to stop opening tasks on a given machine when it reaches the collective task memory limit. Consider something like I want to have celery use all available cores for concurrency as most of the time that will work, but in case it does start to use too much memory collectively, celery should no longer dispatch and wait for memory to go below the threshold. 


Does that make more sense?

**Comment by ashwani-pixis** [comment]
@bdrodes did you find any solutions for issue 1?

**Comment by bdrodes** [comment]
> @bdrodes did you find any solutions for issue 1?

@ashwani-pixis, honestly its been 4 years so I wouldn't recall if I did, but that said, I quickly hit too many limits with Celery if I recall correctly and switched to Nomad, which met all our needs. We never came back to Celery since.

**Comment by ghomem** [comment]
This task is closed as completed. Has it been implemented?

What I understood from the documentation:

> 1. Individual task memory limits. Once a limit is reached, I want celery to send a termination signal to the task. [...]_

It seems that the `--max-memory-per-child` option lets the offending task continue until the end before replacing the worker, which means that a single task could consume all of the server memory without being interrupted.

Ref:
https://docs.celeryq.dev/en/stable/reference/cli.html#cmdoption-celery-worker-max-memory-per-child

> 2. collective task memory limits. An individual task might not consume a lot of memory, but if I have too many running in parallel, the total memory consumption my Out Of Memory my machine. I'd rather not limit the number of cpus that can be used, but instead have celery observe it is over scheduling by monitoring memory. 

I could not find evidence of this being implemented.
